homeassistant/components/enphase_envoy/switch.py

```diff
@@ -0,0 +1,113 @@
+"""Switch platform for Enphase Envoy solar energy monitor."""
+from __future__ import annotations
+
+from collections.abc import Callable, Coroutine
+from dataclasses import dataclass
+import logging
+from typing import Any
+
+from pyenphase import Envoy, EnvoyEnpower
+
+from homeassistant.components.switch import SwitchEntity, SwitchEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import EnphaseUpdateCoordinator
+from .entity import EnvoyBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class EnvoyEnpowerRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[[EnvoyEnpower], bool]
+    turn_on_fn: Callable[[Envoy], Coroutine[Any, Any, dict[str, Any]]]
+    turn_off_fn: Callable[[Envoy], Coroutine[Any, Any, dict[str, Any]]]
+
+
+@dataclass
+class EnvoyEnpowerSwitchEntityDescription(
+    SwitchEntityDescription, EnvoyEnpowerRequiredKeysMixin
+):
+    """Describes an Envoy Enpower switch entity."""
+
+
+ENPOWER_GRID_SWITCH = EnvoyEnpowerSwitchEntityDescription(
+    key="mains_admin_state",
+    translation_key="grid_enabled",
+    value_fn=lambda enpower: enpower.mains_admin_state == "closed",
+    turn_on_fn=lambda envoy: envoy.go_on_grid(),
```

Comment:
Missing await?

Reply:
Nevermind   I see you have it below

Reply:
`turn_on_fn` and `turn_off_fn` get awaited in the platform's async_turn_on/off functions

Reply:
heh, github didn't show your second comment until after I replied 🙃